url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+		JSONObject template = new JSONObject();
+		template.put("type", "INDEX");
+		for(String word: unique_words){
+			JSONObject message = (JSONObject) template.clone();
+			int wordHash = NetworkProperties.hashCode(url);
+			if(wordHash == nodeID){//if word hash matches my id then I should index
+				SearchResponse index = new SearchResponse();
+				index.word = word;
+				index.urls.add(new Url(url,1));
+				indexer.index(index);
+			}else{ // else send to closest node. Might even match exactly
+				Contact closestNode = getClosestNode(wordHash);
+				if(closestNode != null){
+					JSONArray links = new JSONArray();
+					links.add(word);
+					message.put("keyword", word);
+					message.put("link", links);
+					message.put("sender_id", String.valueOf(nodeID));
+					message.put("target_id", String.valueOf(wordHash));
+					InetSocketAddress address = new InetSocketAddress(closestNode.getIp(),NetworkProperties.nPort); 
+					try {
+			            sendQueue.put(jsonToDatagram(message,address));
+			        } catch (InterruptedException e) {
+			            System.err.println("Could not put Datagram on sendQueue in Controller");
+			            e.printStackTrace();
+			        }	
+				}
+			}
+		}
+	}
+	
+	@SuppressWarnings("unchecked")
+	public Vector<SearchResponse> sendSearch(final String[] words) {
+
+		resultResponses = new Vector<SearchResponse>();//we don't know how many results to expect so make a list of them
+		searchResponseWaiter = new Waiter();//spawns a new timer in a map<word,timer> for each search request and when
+		                        // the map is emptied(i.e all searches responded or all pings have been resolved) the thread finishes
+		(new Thread() {// spin a new thread at the end of which the complete set of results is send back to the user
+		public void run() {
+		for (String word: words) {
+			//if my id matches that of the word hash then take straight from my indexer
+			int targetHash = NetworkProperties.hashCode(word);
+			if(targetHash == nodeID){
+				resultResponses.add(indexer.getSearchResponse(word));
+			}else{
+		        //else fire off search queries to appropriate nodes. Assume a node will exist for a keyword
+		        //getting the closest node will have the dual effect of checking if we have direct access to target
+				// or else giving us the closest node's ip.
+		        Contact closestNode = getClosestNode(targetHash);
+		        if(closestNode != null){
+			        JSONObject sendJSON = new JSONObject();
+			        sendJSON.put("type","SEARCH");
+			        sendJSON.put("node_id",Integer.toString(targetHash));
+			        sendJSON.put("word", word);
+			        sendJSON.put("sender_id",Integer.toString(nodeID));
+			        InetSocketAddress sendAddress = new InetSocketAddress(closestNode.getIp(),NetworkProperties.nPort);           				
+			        try {
+			        	sendQueue.put(jsonToDatagram(sendJSON,sendAddress));
+			        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)